execution/gethexec/express_lane_service.go

```diff
@@ -0,0 +1,410 @@
+	// Calculate the next round start time
+	elapsedTime := arrivalTime.Sub(es.initialTimestamp)
+	elapsedRounds := elapsedTime / es.roundDuration
+	nextRoundStart := es.initialTimestamp.Add((elapsedRounds + 1) * es.roundDuration)
+	// Calculate the time to the next round
+	timeToNextRound := nextRoundStart.Sub(arrivalTime)
+	// Check if the arrival timestamp is within AUCTION_CLOSING_DURATION of TIME_TO_NEXT_ROUND
+	return timeToNextRound <= es.auctionClosing
+}
+
+// Sequence express lane submission skips validation of the express lane message itself,
+// as the core validator logic is handled in `validateExpressLaneTx“
+func (es *expressLaneService) sequenceExpressLaneSubmission(
+	ctx context.Context,
+	msg *timeboost.ExpressLaneSubmission,
+) error {
+	// no service lock needed since roundControl is thread-safe
+	control, ok := es.roundControl.Get(msg.Round)
+	if !ok {
+		return timeboost.ErrNoOnchainController
+	}
+
+	es.Lock()
+	defer es.Unlock()
+	// Check if the submission nonce is too low.
+	if msg.SequenceNumber < control.sequence {
+		return timeboost.ErrSequenceNumberTooLow
+	}
+	// Check if a duplicate submission exists already, and reject if so.
+	if _, exists := es.messagesBySequenceNumber[msg.SequenceNumber]; exists {
+		return timeboost.ErrDuplicateSequenceNumber
+	}
+	// Log an informational warning if the message's sequence number is in the future.
+	if msg.SequenceNumber > control.sequence {
+		log.Info("Received express lane submission with future sequence number", "SequenceNumber", msg.SequenceNumber)
+	}
+	// Put into the sequence number map.
+	es.messagesBySequenceNumber[msg.SequenceNumber] = msg
+
+	for {
+		// Get the next message in the sequence.
+		nextMsg, exists := es.messagesBySequenceNumber[control.sequence]
+		if !exists {
+			break
+		}
+		if err := es.transactionPublisher.PublishTimeboostedTransaction(
+			ctx,
+			nextMsg.Transaction,
+			msg.Options,
+		); err != nil {
+			// If the tx failed, clear it from the sequence map.
+			delete(es.messagesBySequenceNumber, msg.SequenceNumber)
+			return err
+		}
+		// Increase the global round sequence number.
+		control.sequence += 1
+	}
+	es.roundControl.Add(msg.Round, control)
+	return nil
+}
+
+func (es *expressLaneService) validateExpressLaneTx(msg *timeboost.ExpressLaneSubmission) error {
+	if msg == nil || msg.Transaction == nil || msg.Signature == nil {
+		return timeboost.ErrMalformedData
+	}
+	if msg.ChainId.Cmp(es.chainConfig.ChainID) != 0 {
+		return errors.Wrapf(timeboost.ErrWrongChainId, "express lane tx chain ID %d does not match current chain ID %d", msg.ChainId, es.chainConfig.ChainID)
+	}
+	if msg.AuctionContractAddress != es.auctionContractAddr {
+		return errors.Wrapf(timeboost.ErrWrongAuctionContract, "msg auction contract address %s does not match sequencer auction contract address %s", msg.AuctionContractAddress, es.auctionContractAddr)
+	}
+	currentRound := timeboost.CurrentRound(es.initialTimestamp, es.roundDuration)
```

Comment:
I've made comments about it elsewhere:
* laways passing around initialTimestamp and roundDuration is really cumbersome
* I think you must take a timestamp when starting to process an incoming message and treat everything from that point as if that's the timestamp

Reply:
This is a very good point, thank you. I fixed this in https://github.com/OffchainLabs/nitro/pull/2841 and I think it makes the code a lot clearer.

I'm going to merge it back in to this PR already so Ganesh and I can keep working against this one.